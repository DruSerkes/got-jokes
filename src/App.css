* {
  background-color: #0F0F0F;
  color: #F0F0F0;
}

.App {
  height: 95vh;
  text-align: center;
  margin-bottom: auto;
  position: relative;
  overflow-y: scroll;
}