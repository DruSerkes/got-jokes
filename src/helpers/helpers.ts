import axios from 'axios';
import { jokeData } from '../types';

const BASE_URL = 'https://v2.jokeapi.dev';

export const wait = (ms: number) => new Promise(res => setTimeout(res, ms));

export const getJokeWithRetry = async (depth = 0): Promise<jokeData> => {
    try {
        const response = await axios.get(`${BASE_URL}/joke/any`);
        const newJoke: jokeData = response.data;
        return newJoke
    } catch (error) {
        if (depth > 7) throw error;
        
        console.log({ error })
        await wait(2 ** depth * 10);
        depth++;
        return getJokeWithRetry(depth);
    }
};


