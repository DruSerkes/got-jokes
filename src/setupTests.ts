// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';

export const testJoke = {
    "error": false,
    "category": 'dark',
    "type": 'single',
    "joke": 'test joke',
    "flags": {
        "nsfw": false,
        "religious": false,
        "political": false,
        "racist": false,
        "sexist": false,
        "explicit": false
    },
    "id": 13,
    "safe": true,
    "lang": 'en',
    'setup': 'test setup',
    'delivery': 'test delivery'
};

export const setJoke = jest.fn(() => console.log('setting joke'));
export const setViewingFavorites = jest.fn(() => console.log('viewing favorites'));
export const setFavorites = jest.fn(() => console.log('setting favorites'));
export const clearFavorites = jest.fn(() => console.log('clearing favorites'));
export const saveFavoriteJoke = jest.fn(() => console.log('saving favorite joke'));