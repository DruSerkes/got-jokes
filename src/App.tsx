import React from 'react';
import { useState, useEffect } from 'react';
import { useLocalStorage } from './hooks/useLocalStorage';
import axios from 'axios';
import Button from 'react-bootstrap/Button';
import { jokeData } from './types';
import { Joke } from './Joke/Joke'
import { SaveJoke } from './SaveJoke/SaveJoke';
import { Favorites } from './Favorites/Favorites';
import { RemoveFavorites } from './RemoveFavorites/RemoveFavorites';
import './App.css';
import Container from 'react-bootstrap/esm/Container';


const BASE_URL = 'https://v2.jokeapi.dev';


// TODO add ability to remove single joke from favorites 
// TODO clean up main App component 
// TODO add tests 
export const App = () => {
  const [joke, setJoke] = useState<jokeData | undefined>(undefined);
  const [favorites, setFavorites] = useLocalStorage('favorites');
  const [isLoading, setIsLoading] = useState(true);
  const [viewingFavorites, setViewingFavorites] = useState(false);
  const handleShowFavorites = () => setViewingFavorites(!viewingFavorites);
  const clearFavorites = () => {
    setFavorites([])
    setViewingFavorites(false);
  };
  const handleGetJoke = () => {
    setJoke(undefined);
    setViewingFavorites(false);
  };
  const saveFavoriteJoke = () => {
    if (!joke) return;
    const newFavorites: jokeData[] = [joke, ...favorites];
    setFavorites(newFavorites);
    setJoke(undefined)
  };

  // Get a joke if you don't have one
  useEffect(() => {
    const getJoke = async () => {
      if (!joke) {
        setIsLoading(() => true);
        const response = await axios.get(`${BASE_URL}/joke/any`);
        const newJoke: jokeData = response.data;
        console.log(newJoke);
        setJoke(newJoke);
        setIsLoading(() => false);
      }
    }
    getJoke();
  }, [joke]);

  return (
    <Container className="App">
      <h1 className='h1 my-4'>Got Jokes</h1>

      <Button
        variant="primary"
        onClick={handleGetJoke}
        disabled={isLoading ? true : false}
        className='my-4 mx-2'>
        Make Me Laugh!
        </Button>

      <SaveJoke joke={joke} viewingFavorites={viewingFavorites} saveFavoriteJoke={saveFavoriteJoke} />

      <Button
        variant="secondary"
        onClick={handleShowFavorites}
        className='my-4 mx-2'>
        {!viewingFavorites ? 'Show my Favorites' : 'Hide my Favorites'}
      </Button>

      <RemoveFavorites clearFavorites={clearFavorites} haveFavorites={!!favorites.length} viewingFavorites={viewingFavorites} />

      {!viewingFavorites && (<Joke isLoading={isLoading} joke={joke} />)}

      <Favorites favorites={favorites} viewingFavorites={viewingFavorites} />

    </Container>
  );
}