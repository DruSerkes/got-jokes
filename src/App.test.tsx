import * as React from 'react';
import { render, screen, act } from '@testing-library/react';
import { App } from './App';
import { wait } from './helpers/helpers';
import { testJoke } from './setupTests';
import * as helpers from './helpers/helpers';

describe('App tests', () => {


  it('renders without breaking and matches snapshot', async () => {
    const spy = jest.spyOn(helpers, 'getJokeWithRetry').mockResolvedValue(testJoke);

    await act(async () => {
      const { asFragment } = render(<App />);
      await wait(800)
      expect(asFragment()).toMatchSnapshot();
    })

  });

  it('renders the test joke', async () => {
    const spy = jest.spyOn(helpers, 'getJokeWithRetry').mockResolvedValue(testJoke);

    await act(async () => {
      render(<App />);
      await wait(800);
    });

    const { getByText } = screen;
    const jokeText = getByText(testJoke.joke);
    expect(jokeText).toBeInTheDocument();
  });

});


