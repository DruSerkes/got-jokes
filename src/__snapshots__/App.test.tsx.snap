// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests renders without breaking and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App container"
  >
    <h1
      class="h1 my-4"
    >
      Got Jokes
    </h1>
    <div
      aria-label="Main App Buttons"
      class="Buttons btn-group"
      role="group"
    >
      <button
        class="my-4 btn btn-primary"
        type="button"
      >
        Make Me Laugh!
      </button>
      <button
        class="my-4 btn btn-info"
        type="button"
      >
        Save to Favorites
      </button>
      <button
        class="my-4 btn btn-secondary"
        type="button"
      >
        Show my Favorites
      </button>
    </div>
    <div
      class="Joke container"
    >
      <div
        class="my-3 row"
      >
        <div
          class="col"
        >
          test joke
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
